File: 113/113_version1.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode(int x) { val = x; }
 * }
 */
class Solution {
    private List<List<Integer>> res = new ArrayList<>();
    private int target;
    public List<List<Integer>> pathSum(TreeNode root, int sum) {
        this.target = sum;
        dfs(root,0,new ArrayList<>());
        return res;
    }
    
    private void dfs(TreeNode node,int currSum,List<Integer> pathRecord){
        if(node!=null){
            pathRecord.add(node.val);
            if(node.left==null && node.right==null){
                if(currSum+node.val==target){
                    res.add(new ArrayList<>(pathRecord));
                }
            } else {
                dfs(node.left,currSum+node.val,pathRecord);
                dfs(node.right,currSum+node.val,pathRecord);
            }
            pathRecord.remove(pathRecord.size()-1);
        }
    }
}

File: 113/113_version1.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Solution:
    def pathSum(self, root: TreeNode, sum: int) -> List[List[int]]:
        self.target = sum
        self.res = []
        self.dfs(root,0,[])
        return self.res
    def dfs(self,node,curr_sum,path_record):
        if node:
            path_record.append(node.val)
            if not node.right and not node.left:
                if curr_sum+node.val == self.target: # is leaf node
                    self.res.append([i for i in path_record])
            else:
                self.dfs(node.left,curr_sum+node.val,path_record)
                self.dfs(node.right,curr_sum+node.val,path_record)
            path_record.pop()
